### DISCUSSION TITLE
websocket

### DISCUSSION THREAD

**Discussion by Cyrila54** [vercel/next.js#52801] [Help]
### Summary

hello, I have a problem, I deployed my backend (nodejs) and my frontend (next/react) on vercel, no worries, the problem is that the websocket part (socket.io) does not work and I do not understand why because it works very well in localhost.

Thank you very much for your help. attached my console error.

cyril

### Additional information

```js
521-b75fc05d5054e918.js:1     GET https://skyjo-backend.vercel.app/socket.io/?EIO=4&transport=polling&t=Obb24tX 404
create @ 521-b75fc05d5054e918.js:1
V @ 521-b75fc05d5054e918.js:1
request @ 521-b75fc05d5054e918.js:1
doPoll @ 521-b75fc05d5054e918.js:1
poll @ 521-b75fc05d5054e918.js:1
doOpen @ 521-b75fc05d5054e918.js:1
open @ 521-b75fc05d5054e918.js:1
open @ 521-b75fc05d5054e918.js:1
G @ 521-b75fc05d5054e918.js:1
open @ 521-b75fc05d5054e918.js:1
(anonyme) @ 521-b75fc05d5054e918.js:1
setTimeout (asynchrone)
reconnect @ 521-b75fc05d5054e918.js:1
(anonyme) @ 521-b75fc05d5054e918.js:1
i @ 521-b75fc05d5054e918.js:1
k.emit @ 521-b75fc05d5054e918.js:1
onError @ 521-b75fc05d5054e918.js:1
k.emit @ 521-b75fc05d5054e918.js:1
onError @ 521-b75fc05d5054e918.js:1
(anonyme) @ 521-b75fc05d5054e918.js:1
k.emit @ 521-b75fc05d5054e918.js:1
onError @ 521-b75fc05d5054e918.js:1
(anonyme) @ 521-b75fc05d5054e918.js:1
setTimeout (asynchrone)
r.onreadystatechange @ 521-b75fc05d5054e918.js:1
XMLHttpRequest.send (asynchrone)
create @ 521-b75fc05d5054e918.js:1
V @ 521-b75fc05d5054e918.js:1
request @ 521-b75fc05d5054e918.js:1
doPoll @ 521-b75fc05d5054e918.js:1
poll @ 521-b75fc05d5054e918.js:1
doOpen @ 521-b75fc05d5054e918.js:1
open @ 521-b75fc05d5054e918.js:1
open @ 521-b75fc05d5054e918.js:1
G @ 521-b75fc05d5054e918.js:1
open @ 521-b75fc05d5054e918.js:1
(anonyme) @ 521-b75fc05d5054e918.js:1
```

**Comment by icyJoseph** [comment]
How's the backend deployed? It might just be that it is deployed as a serverless backend, and the socket dies out after a timeout time.

**Comment by Cyrila54** [comment]
thank you for your answer @icyJoseph 
I deployed it on vercel in a "traditional" way. I'm a young developer so I'm not sure I fully understand the question but if you help me understand the question better, I can tell you.

I simply deployed via vercel by making new project. my backend is in a separate repository. Is there something to check on the backend specifically?

**Reply by icyJoseph** [reply]
What kind of tech stack is the backend using? Is it using a project template? 

You might want to consider other kinds of hosting such as: https://render.com/ or https://railway.app/ if you need the app to never "sleep".

Serverless apps are only awake to serve a request and then they spin down. Which would explain why the socket is not working.

**Reply by Cyrila54** [reply]
I tried other platforms (netlify, heroku,...) but each one has its problems which are added to the previous ones.

I understand your explanation about serverless applications (and thank you) but it's still supposed to work at least a little if it goes to sleep afterwards.
On my side, it doesn't work at all once the deployment is done on vercel. (not even at the beginning)

regarding your question i am using node js and express

I admit to being a little overwhelmed by this problem.

**Reply by icyJoseph** [reply]
Aha but that's ok. Perhaps I've jumped ahead with the issue.

What is exactly not working? And how do you see this? Are there any logs?

**Reply by icyJoseph** [reply]
Also make sure to have the same Node.js version on your deployment as you do locally.

**Reply by Cyrila54** [reply]
I use websockets to return cards to all players at the same time, which works very well locally as I said, but which no longer works once deployed.

here is a log error (repeating). I did not read your last post, I see that you just sent it to me I will look.

Also make sure to have the same Node.js version on your deployment as you do locally : ok i will look

```
Time
July 17 20:45:02.83 GMT+02:00
Status Code
404
Host
skyjo-backend.vercel.app
Request ID
xhxbw-1689619502829-790271a83ad7
Request User Agent
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36
Level
Warning

Cache
MISS
Type
Serverless Function

Function
/app.js
Location
Washington, D.C., USA (iad1)
Runtime
Node.js 18.x

Execution Duration / Limit
3ms / 10s
Cold Boot Duration
0ms
Memory Used / Limit
127 MB / 1024 MB
[0mGET /socket.io/?EIO=4&t=ObbAvPi&transport=polling [33m404 [0m0.443 ms - 149[0m
```

**Reply by Cyrila54** [reply]
node -v 
v18.15.0

**Reply by icyJoseph** [reply]
So your frontend sees a 404? 

Sorry I just can't stop thinking that this is a problem with the serverless deployment. Heroku and Netlify might also use serverless deployments... 

Is this an express app?

**Reply by Cyrila54** [reply]
yes I had written to you a little higher.

**Reply by icyJoseph** [reply]
Alright, yeah sorry GitHub collapsed the replies.

So express app, that's totally reachable at https://skyjo-backend.vercel.app but not at the path of the socket... Well I don't know, I really think the function spins up and shutdown

Are you doing something like this? https://vercel.com/guides/using-express-with-vercel#standalone-express

**Reply by Cyrila54** [reply]
Hello,
yes there are some similar points indeed. Thanks for trying to help me though @icyJoseph

**Comment by AlessandroFare** [comment]
Vercel serverless functions are not meant to hold long-lived WebSockets. Each invocation spins up, handles a request, and exits, so a Socket.IO server on the same deployment will drop or behave oddly under load.

Common fixes:
1. Split deploy: Next on Vercel, socket server on Fly/Railway (you lose "all on Vercel").
2. Edge realtime: Next on Vercel; chat on Cloudflare Worker + one Durable Object per room (FluxyChat pattern).
3. Managed vendor: Pusher/Ably (second bill, less ops).

If you need rooms + history + reconnect without a socket VM, (2) is what I'd pick. What are you building, chat threads or fire-and-forget alerts?
